domain/annotation/state/query.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/annotations"
+	annotationerrors "github.com/juju/juju/domain/annotation/errors"
+)
+
+// getAnnotationQueryForID provides a query for the given id, based on pre-computed queries for
+// GetAnnotations for different kinds of ids. We keep these static (avoiding dynamically generating
+// table names and fields) to keep things safe.
+func getAnnotationQueryForID(id annotations.ID) (string, error) {
+	var query string
+
+	if id.Kind == annotations.KindModel {
+		query = `SELECT (key, value) AS (&Annotation.*) from annotation_model`
+	} else {
+		kindName, err := kindNameFromID(id)
+		if err != nil {
+			return "", errors.Trace(err)
+		}
+		query = fmt.Sprintf(`
+SELECT (key, value) AS (&Annotation.*)
+FROM annotation_%s
+WHERE uuid = $M.uuid`, kindName)
+	}
+
+	return query, nil
+}
+
+// setAnnotationQueryForID provides a query for the given id, based on pre-computed queries for
+// SetAnnotations for different kinds of ids. We keep these static (avoiding dynamically generating
+// table names and fields) to keep things safe.
+func setAnnotationQueryForID(id annotations.ID) (string, error) {
+	var query string
+
+	if id.Kind == annotations.KindModel {
+		query = `
+INSERT INTO annotation_model (key, value)
+VALUES ($M.key, $M.value)
+	ON CONFLICT(key) DO UPDATE SET value=$M.value`
+	} else {
+		kindName, err := kindNameFromID(id)
+		if err != nil {
+			return "", errors.Trace(err)
+		}
+		query = fmt.Sprintf(`
+INSERT INTO annotation_%s (uuid, key, value)
+VALUES ($M.uuid, $M.key, $M.value)
+	ON CONFLICT(uuid, key) DO UPDATE SET value=$M.value`, kindName)
+	}
+
+	return query, nil
+}
+
+// deleteAnnotationQueryForID provides a query for the given id, based on pre-computed queries for
+// SetAnnotations for different kinds of ids. We keep these static (avoiding dynamically generating
+// table names and fields) to keep things safe.
+func deleteAnnotationsQueryForID(id annotations.ID, toRemoveBindings string) (string, error) {
+	var query string
+	if id.Kind == annotations.KindModel {
+		query = fmt.Sprintf(`DELETE FROM annotation_model WHERE key IN (%s)`, toRemoveBindings)
+	} else {
+		kindName, err := kindNameFromID(id)
+		if err != nil {
+			return "", errors.Trace(err)
+		}
+		query = fmt.Sprintf(`
+DELETE FROM annotation_%s
+WHERE uuid = $M.uuid AND key IN (%s)`, kindName, toRemoveBindings)
```

Comment:
You can use the SQLair syntax `$S[:]` to pass a slice here. See [here](https://github.com/juju/juju/blob/e0fbecd1da5b1cd0cc34b606c9b003eda5167c97/domain/modelconfig/state/state.go#L63) for an example. In that example I used the `sqlair.S` type, though you can use any named type and in a few other places I created a new type to use, e.g. in this case you could use `type Keys []string` and pass that as a type argument to `sqlair.Prepare` then put `$Keys[:]` in the query.